File: cpp/includes/cloud/geolocation/geolocation.cpp
#include "geolocation.hpp"
namespace rapp {
namespace cloud {

geolocation::geolocation(
             const std::string ipaddr,
             const std::string engine,
             std::function<void(std::string,
                                std::string,
                                std::string,
                                float,
                                float,
                                std::string,
                                std::string)> callback           )

: http_request("POST /hop/geolocation HTTP/1.1\r\n"), 
  delegate_(callback)
{

    std::string ss;
    json json_doc = {{"ipaddr", ipaddr},
                    {"engine", engine}};
    ss = json_doc.dump(4);
    http_request::add_content("json", ss, false); 
    http_request::close();
 
}

void geolocation::deserialise(std::string json) const
{
   
    std::string city;
    std::string country; 
    std::string country_code; 
    float latitude;
    float longtitude;
    std::string timezone; 
    std::string zip; 

    std::string  current_class = "geolocation";

    auto json_f = json::parse(json);

    if (!json.empty()) {

        // Get "city" from json
        rapp::misc::get_value_from_json( current_class, "city", json_f, city); 

        // Get "country" from json
        rapp::misc::get_value_from_json( current_class, "country", json_f, country);        

        // Get "country_code" from json
        rapp::misc::get_value_from_json( current_class, "country_code", json_f, country_code); 

        // Get "latitude" from json 
        rapp::misc::get_value_from_json( current_class, "latitude", json_f, latitude);
        
        // Get "longtitude" from json
        rapp::misc::get_value_from_json( current_class, "longtitude", json_f, longtitude);        

        // Get "timezone" from json
        rapp::misc::get_value_from_json( current_class, "timezone", json_f, timezone); 

        // Get "zip" from json
        rapp::misc::get_value_from_json( current_class, "zip", json_f, zip);
        
        // Get error from json
        std::string error;

        rapp::misc::get_value_from_json( current_class, "error", json_f, error);
        if (!error.empty()) {
            std::cerr << "cognitive_test_selector JSON error: " << error <<std::endl;
        }
    }
    else {
        throw std::runtime_error("json is empty");
    }

    delegate_(city,
              country,
              country_code,
              latitude,
              longtitude,
              timezone,
              zip);
}


}
}

File: js/lib/cloud/geolocation.js
#!/usr/bin/env node

var path = require('path');
var __cloudDir = path.join(__dirname);
var RAPPCloud = require(path.join(__cloudDir, 'RAPPCloud.js'));

/**
 * @fileOverview Prototype the RAPPCloud Service Method.
 * 
 * @class geolocation
 * @memberof RAPPCloud
 * @description Asynchronous Service which will request the geolocation RAPP platform service
 * @version 1
 * @author Lazaros Penteridis <lp@ortelio.co.uk>
 * @param ipaddr: The machine's ipaddr
 * @param engine: Engine to use. Defaults to 'ip-api' (Currently the only supported).
 * @param callback is the function that will receive the result
 */
RAPPCloud.prototype.geolocation = function(
                                            ipaddr,
                                            engine,
                                            callback
                                          )
{
    var cloud = this;
    var _delegate = callback;
    var request = cloud.determine_protocol();

    var body_obj = {};
    body_obj.ipaddr = cloud.escape_string(ipaddr);
    body_obj.engine = cloud.escape_string(engine);
    var body_json = JSON.stringify(body_obj);
    
    request.post({
        headers: {
			'Accept-Token' : cloud.token,
			'Connection' : 'close'
			},
        url: cloud.cloud_url + '/hop/geolocation/ ',
        body: "json=" + body_json
    },
    function(error, response, json) 
    {
        if (!error && response.statusCode == 200)
            handle_reply(json);
        else if (error)
            error_handler(error);
        else if (response.statusCode != 200)
            console.log('Error: ' + response.statusCode);
    });
    
    function handle_reply(json)
    {
		var json_obj;
		try {
			json_obj = JSON.parse(json);
			// JSON reply is: { city: '', country: '', country_code: '', 
            // latitude: 0.0, longtitude: 0.0, region: '', timezone: '',
            // zip: '', error: '' }	
            //
		    // city: (String): The city.
		    // country (String): The country.
		    // country_code (String): The country code.
		    // latitude: (Float): The latitude.
		    // longtitude (Float): The longtitude.
		    // timezone (String): The timezone.
		    // zip (String): The zip postal code.
		    // error (String): Error message, if one occures.
		
			if(json_obj.error)  // Check for Errors  
				console.log('geolocation JSON error: ' + json_obj.error);
	        else 
	        	_delegate(
                            json_obj.city,
                            json_obj.country,
                            json_obj.country_code,
                            json_obj.latitude,
                            json_obj.longtitude,
                            json_obj.timezone,
                            json_obj.zip
                         );
		} catch (e) {
			console.log('geolocation::handle_reply Error parsing: ');
			return console.error(e);
		}
	}
	
	function error_handler(error) {
		return console.error(error);
	}   
};

/// Export
module.exports = RAPPCloud.geolocation;
